shared/src/commonMain/kotlin/io/customer/shared/tracking/queue/QueueDispatcher.kt

```diff
@@ -0,0 +1,154 @@
+import io.customer.shared.tracking.constant.QueueTaskStatus
+import io.customer.shared.util.DateTimeUtil
+import io.customer.shared.util.Logger
+import io.customer.shared.work.*
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.datetime.Instant
+import kotlin.time.DurationUnit
+
+/**
+ * Class responsible for running queue operations. This class only receives requests to validate the
+ * queue and is responsible for following:
+ *
+ * - whether to run the queue now or wait (except when specifically asked to run).
+ * - maintain lock to avoid sending same tasks multiple times.
+ * - validate and update task status related to queue operations.
+ * - schedule timer for next queue request.
+ */
+internal interface QueueDispatcher {
+    suspend fun checkForPendingTasks(source: QueueTriggerSource)
+    suspend fun sendAllPending()
+}
+
+internal class QueueDispatcherImpl(
+    private val logger: Logger,
+    private val dateTimeUtil: DateTimeUtil,
+    override val jobExecutor: JobExecutor,
+    private val backgroundQueueConfig: BackgroundQueueConfig,
+    private val trackingTaskQueryHelper: TrackingTaskQueryHelper,
+    private val queueRunner: QueueRunner,
+    private val queueTimer: QueueTimer,
+) : QueueDispatcher, JobDispatcher {
+    private val mutex = Mutex()
+
+    private suspend fun acquireLock() {
+        logger.debug("Acquiring lock for queue")
+        kotlin.runCatching {
+            mutex.lock(owner = this)
+        }.onFailure { ex ->
+            logger.error("Mutex locking failed, reason: $ex")
+        }
+    }
+
+    private fun releaseLock() {
+        logger.debug("Releasing lock for queue")
+        kotlin.runCatching {
+            mutex.unlock()
+        }.onFailure { ex ->
+            logger.error("Mutex unlocking failed, reason: $ex")
+        }
+    }
+
+    override fun onCoroutineFailed(exception: Throwable) {
+        releaseLock()
+    }
+
+    override suspend fun checkForPendingTasks(source: QueueTriggerSource) {
+        processBatchTasks(isRecursive = false) { getTasksToBatch() }
+    }
+
+    override suspend fun sendAllPending() {
+        processBatchTasks(isRecursive = true) {
+            trackingTaskQueryHelper.selectAllPendingTasks().getOrNull()
+        }
+    }
+
+    /**
+     * Since we have imposed a limit of maximum number of tasks, [isRecursive] allows the caller
+     * to send multiple batch events. If true, it will keep calling [getTasks] after every
+     * successful execution until an empty list of tasks is provided.
+     */
+    private fun processBatchTasks(
+        isRecursive: Boolean,
+        getTasks: suspend () -> List<TrackingTask>?,
+    ) = runOnBackground(
+        onFailure = { releaseLock() },
+    ) {
+        if (mutex.isLocked) return@runOnBackground
+
+        acquireLock()
+        kotlin.runCatching {
+            queueTimer.cancel()
+        }.onFailure { ex ->
+            logger.error("Cannot cancel scheduled timer error: ${ex.message}")
+        }
```

Comment:
Yeah, agreed. From the [docs](https://developer.android.com/reference/android/os/CountDownTimer#cancel()), it doesn't look like throwing exception for any case so it can be avoided, removing this.